File: webots/controllers/pusher_common.py
import math
import numpy as np
import random
import pusher_globals as g

def changeState(newState):
    print(f'Changed to state "{newState}"')
    g.state = newState

########## INITIALIZE ##########
def initRobot():
    # Initialize infra red sensors
    for i in range(8):
        g.irs.append(g.robot.getDevice('ir'+str(i)))
        g.irs[i].enable(g.TIME_STEP)

    # Initialize cameras
    for i in range(4):
        g.cams.append(g.robot.getDevice('cam'+str(i)))
        g.cams[i].enable(g.TIME_STEP)

    # Initialize motors
    g.leftMotor.setPosition(float('inf'))
    g.rightMotor.setPosition(float('inf'))
    g.leftMotor.setVelocity(0.0)
    g.rightMotor.setVelocity(0.0)

######## MOVEMENT CONTROLLERS ###########
def dirToVec(direction):
    '''
    Given the direction, convert to vector.
    Y axis is forward, X axis is to the right.
    '''
    return np.array([math.sin(math.radians(direction)), math.cos(math.radians(direction))])

def vecToMotor(moveVector):
    '''
    Given the movement vector, set the motor speeds.
    Y axis is forward, X axis is to the right.
    '''
    norm = np.linalg.norm(moveVector)
    ang = 0

    # Allow swap front every X milliseconds to avoid getting stuck
    if g.worldTime % vecToMotor.swapTime == 0:
        vecToMotor.moveForward =  moveVector[1] >= 0

    if vecToMotor.moveForward:
        ang = math.atan2(moveVector[0], moveVector[1])
    else:
        ang = math.atan2(-moveVector[0], moveVector[1])

    # Calculate power
    power = np.array([math.cos(ang) - math.sin(ang), math.cos(ang) + math.sin(ang)])
    power *= norm

    # Normalize power
    if abs(power[0]) > 1.0:
        power /= abs(power[0])
    if abs(power[1]) > 1.0:
        power /= abs(power[1])

    # Apply power
    g.leftMotor.setVelocity(g.MAX_SPEED*power[0])
    g.rightMotor.setVelocity(g.MAX_SPEED*power[1])

vecToMotor.moveForward = 1
vecToMotor.swapTime = 512

########## UTILS ##########
def angleDistance(a0, a1):
    dist = (a1-a0) if a1 > a0 else (a0-a1)
    if dist > 180:
        return 360 - dist
    return dist

########## SENSING ##########
def canSeeObject():
    return cameraCheckColor(g.objectColor)
def canSeeGoal():
    return cameraCheckColor(g.goalColor)
def cameraCheckColor(color):
    for y in range(g.IMAGE_MAX_ROW, -1, -1):# From IMAGE_MAX_ROW to 0
        for i in range(4):
            for x in range(g.IMAGE_SIZE):
                r = g.cams[i].imageGetRed(g.cams[i].getImage(), g.IMAGE_SIZE, x, y)
                gr = g.cams[i].imageGetGreen(g.cams[i].getImage(), g.IMAGE_SIZE, x, y)
                b = g.cams[i].imageGetBlue(g.cams[i].getImage(), g.IMAGE_SIZE, x, y)
                if color.check(r, gr, b):
                    #file = 'recognized.png'
                    #cam.saveImage(file, 100)
                    #print(f'r:{r} g:{g} b{b} -> image saved {file}')
                    return True
    return False

def colorDirectionScan(color, rowRange, checkFree):
    '''
    Color direction scan algorithm

    Input: color, rowRange, checkFree
    color -> Color that is search for
    rowRange -> Sequence of rows to check in the images
    checkFree -> If should ignore color if there is robot color below

    Return: found, direction, freeSpace
    found -> If found a row of interest
    direction -> Direction from -180.0 to 180.0 (0.0 is in front of the robot)
    freeSpace -> Size of the free space in pixels

    The algorithm will scan the image following the rowRange searching for the first row
    of color pixels (if checkFree, the pixel should not have robot pixel below). After
    finding the row of interest, the algorithm calculates the angle from the robot to
    the color [-180, 180] (0 being front, -90 being left, 90 being right). Only largest
    interval of color pixels in the row is used to calculate the angle.
    '''
    images = [g.cams[0].getImage(), g.cams[1].getImage(), g.cams[2].getImage(), g.cams[3].getImage()]
    for y in rowRange:# From IMAGE_MAX_ROW to 0
        for i in range(4):
            for x in range(g.IMAGE_SIZE):
                r = g.cams[i].imageGetRed(images[i], g.IMAGE_SIZE, x, y)
                gr = g.cams[i].imageGetGreen(images[i], g.IMAGE_SIZE, x, y)
                b = g.cams[i].imageGetBlue(images[i], g.IMAGE_SIZE, x, y)

                # Check if pixel is object pixel
                if color.check(r, gr, b):
                    # Find intervals of object pixels in this row without robot pixels below
                    intervals = []
                    start = 0
                    end = 0
                    for I in range(4):
                        for X in range(g.IMAGE_SIZE):
                            r = g.cams[I].imageGetRed(images[I], g.IMAGE_SIZE, X, y)
                            gr = g.cams[I].imageGetGreen(images[I], g.IMAGE_SIZE, X, y)
                            b = g.cams[I].imageGetBlue(images[I], g.IMAGE_SIZE, X, y)
                            rBelow = g.cams[I].imageGetRed(images[I], g.IMAGE_SIZE, X, y+1)
                            gBelow = g.cams[I].imageGetGreen(images[I], g.IMAGE_SIZE, X, y+1)
                            bBelow = g.cams[I].imageGetBlue(images[I], g.IMAGE_SIZE, X, y+1)

                            if not color.check(r, gr, b) or (checkFree and g.robotColor.check(rBelow, gBelow, bBelow)):
                                if start != end:
                                    intervals.append([start, end])
                                start = X + I*g.IMAGE_SIZE
                                end = X + I*g.IMAGE_SIZE
                            else:
                                end = X + I*g.IMAGE_SIZE
                    if start != end:
                        intervals.append([start, end])

                    # Merge intervals (wrapping around)
                    idxBeginInt = -1
                    idxEndInt = -1
                    for i, interval in enumerate(intervals):
                        if interval[0] == 0:
                            idxBeginInt = i
                        if interval[1] == 4*g.IMAGE_SIZE-1:
                            idxEndInt = i
                    if idxBeginInt != -1 and idxEndInt != -1:
                        intervals[idxBeginInt][0] = intervals[idxEndInt][0]
                        intervals = intervals[:idxEndInt] + intervals[idxEndInt+1 :]

                    # Find largest interval
                    idxLargest = 0
                    sizeLargest = 0
                    for i, interval in enumerate(intervals):
                        size = interval[1] - interval[0]
                        if interval[0] > interval[1]:
                            size = interval[1] + (g.IMAGE_SIZE*4-interval[0])
                        if size >= sizeLargest:
                            idxLargest = i
                            sizeLargest = size

                    # Calculate direction
                    if sizeLargest == 0:
                        continue
                    else:
                        interval = intervals[idxLargest]
                        # Calculate mean pixel position
                        pixelPos = (interval[0] + sizeLargest/2)%(g.IMAGE_SIZE*4)
                        # Convert to [0,1]
                        meanPos = pixelPos/(g.IMAGE_SIZE*4.0)
                        # Convert to [-2, 2]
                        meanPos = (meanPos*4 - 0.5)
                        if meanPos > 2:
                            meanPos = -2+(meanPos-2)
                        # Calculate direction
                        direction = meanPos*90
                        return True, direction, sizeLargest
    return False, 0, 0

def freeDirectionToObject():
    '''
    Free direction to object algorithm

    Return: isFree, direction, freeSpace
    isFree -> If there is free space to push the object
    direction -> Direction from -180.0 to 180.0 (0.0 is in front of the robot)
    freeSpace -> Size of the free space in pixels

    The algorithm will scan the image from bottom to top searching for the first row
    of object pixels that does not contain robot pixels below. After finding the row of
    interest, the algorithm calculates the angle from the robot to the object [-180, 180]
    (0 being front, -90 being left, 90 being right). Only largest interval of object pixels
    in the row is used to calculate the angle.
    '''
    return colorDirectionScan(color=g.objectColor, rowRange=range(g.IMAGE_MAX_ROW, -1, -1), checkFree=True)


def directionToObject():
    '''
    Direction to object algorithm

    Return: found, direction
    found -> If found the object
    direction -> Estimated object direction

    The algorithm will scan the image from top to bottom searching the first row
    with object pixels. The average object pixel position in that row is converted
    to direction angle and is returned
    '''
    found, direction, _ = colorDirectionScan(color=g.objectColor, rowRange=range(g.IMAGE_SIZE), checkFree=False)
    return found, direction

def directionToGoal():
    '''
    Direction to goal algorithm

    Return: found, direction
    found -> If found the object
    direction -> Estimated object direction

    Similar to directionToObject
    '''
    found, direction, _ = colorDirectionScan(color=g.goalColor, rowRange=range(g.IMAGE_SIZE), checkFree=False)
    return found, direction

def distanceToObject():
    return distanceToColor(g.objectColor)
def distanceToGoal():
    return distanceToColor(g.goalColor)

def distanceToColor(color):
    '''
    Distance to algorithm

    Return: distance
    distance -> Number of row between top of the image and first object pixel

    The algorithm will scan the image from top to bottom searching for the first row
    that contains object pixel. If the object is visible in the whole image, 0 is
    returned. If the object is not in the image, IMAGE_SIZE is returned.
    '''
    for y in range(g.IMAGE_SIZE):# From 0 to IMAGE_SIZE-1
        for i in range(4):
            for x in range(g.IMAGE_SIZE):
                r = g.cams[i].imageGetRed(g.cams[i].getImage(), g.IMAGE_SIZE, x, y)
                gr = g.cams[i].imageGetGreen(g.cams[i].getImage(), g.IMAGE_SIZE, x, y)
                b = g.cams[i].imageGetBlue(g.cams[i].getImage(), g.IMAGE_SIZE, x, y)

                # Check if pixel is object pixel
                if color.check(r, gr, b):
                    return y
    return g.IMAGE_SIZE


########## STATES ##########
def randomWalk():
    wallDistParam = 400
    if randomWalk.dirFront:
        nearWall = (g.irs[0].getValue() < wallDistParam) \
            or (g.irs[1].getValue() < wallDistParam)     \
            or (g.irs[7].getValue() < wallDistParam)
    else:
        nearWall = (g.irs[4].getValue() < wallDistParam) \
            or (g.irs[5].getValue() < wallDistParam)     \
            or (g.irs[3].getValue() < wallDistParam)

    # Check if should stop random walk
    if canSeeObject() and canSeeGoal():
        changeState(g.State.APPROACH_OBJECT)
        randomWalk.dirFront = not randomWalk.dirFront

    elif not nearWall and not randomWalk.RW_turning:
        # Move in straight line
        randomWalk.counter = 0
        if randomWalk.dirFront:
            g.leftMotor.setVelocity(g.MAX_SPEED)
            g.rightMotor.setVelocity(g.MAX_SPEED)
        else:
            g.leftMotor.setVelocity(-g.MAX_SPEED)
            g.rightMotor.setVelocity(-g.MAX_SPEED)
    else:
        randomWalk.RW_turning = True
        if randomWalk.counter < randomWalk.turnTimer:
            # Turning
            g.leftMotor.setVelocity(-g.MAX_SPEED)
            g.rightMotor.setVelocity(g.MAX_SPEED)
            randomWalk.counter += g.TIME_STEP
        else:
            # Finished turning
            randomWalk.RW_turning = False
            randomWalk.turnTimer = random.randint(512, 512*3)
randomWalk.dirFront = True
randomWalk.RW_turning = False
randomWalk.counter = 0
randomWalk.turnTimer = 0

def approachObject(checkGoal = False):
    minDist = 5# Minimum distance to the object to change state
    minAngle = 10# The front angle interval is [-minAngle, minAngle]

    objectIsVisible, direction = directionToObject()
    goalIsVisible = canSeeGoal()
    if objectIsVisible and ((not checkGoal) or goalIsVisible):
        vecToMotor(dirToVec(direction))
    else:
        changeState(g.State.RANDOM_WALK)
        return

    # Check if arrived
    dist = distanceToObject()
    isInFront = (direction < minAngle and direction > -minAngle) \
        or (direction > -math.pi+minAngle or direction > math.pi-minAngle)
    if dist < minDist and isInFront:
        if goalIsVisible:
            # Close enough to object and is looking straight at it
            changeState(g.State.MOVE_AROUND_OBJECT)
        else:
            changeState(g.State.PUSH_OBJECT)

def moveAroundObject(clockwise = True):
    #----- Parameters -----#
    moveVec = np.array([0, 1]) # Robot move vector (X is to the right, Y is forward)
    minDist = 300

    #----- Input -----#
    distToObject = distanceToObject()
    visible, dirToObject = directionToObject()

    idxF = 0 if clockwise else 4
    if dirToObject < 0:
        # If robot moving "backward", front sensor is behind
        if idxF == 4:
            idxF = 0
        else:
            idxF = 4

    irF = g.irs[idxF].getValue()# IR front
    if clockwise:
        irS = g.irs[idxF+2].getValue()# IR right
    else:
        irS = g.irs[(idxF-2+len(g.irs))%len(g.irs)].getValue()# IR left

    #----- Check lost object -----#
    if not visible:
        changeState(g.State.RANDOM_WALK)

    #----- Check should push -----#
    if g.worldTime % 1024 == 0:
        # Check every 2 seconds if should push the object
        if not canSeeGoal():
            isFree, dirToFree, _ = freeDirectionToObject()
            if isFree and angleDistance(dirToFree, dirToObject) < 10:
                changeState(g.State.PUSH_OBJECT)

    #----- Force field -----#
    objVec = dirToVec(dirToObject)

    # Force to move around the object
    moveVec = np.array([-objVec[1], objVec[0]])
    if not clockwise:
        moveVec *= -1

    # Force to keep distance from object
    if irS > g.IR_SENSOR_LIMIT:
        moveVec += objVec
    elif irS < minDist:
        moveVec -= objVec

    # Force to deviate from obstacles
    if irF < minDist:
        moveVec = np.array([-1,0])

    #----- Output - move -----#
    vecToMotor(moveVec)

    #----- Timeout -----#
    # If timer reached zero
    if moveAroundObject.timer == 0:
        changeState(g.State.RANDOM_WALK)
        moveAroundObject.timer = moveAroundObject.timeout
    moveAroundObject.timer -= g.TIME_STEP
moveAroundObject.timeout = 1024*2*60# Timeout 2min
moveAroundObject.timer = moveAroundObject.timeout

def pushObject():
    isFree, direction, _ = freeDirectionToObject()
    dist = distanceToObject()

    if isFree:
        if dist == 0:
            # If pushing the object, push slowly
            vecToMotor(dirToVec(direction)*0.3)
        else:
            # If approaching the object, move fast
            vecToMotor(dirToVec(direction))
    else:
        #print(f'No free space')
        changeState(g.State.MOVE_AROUND_OBJECT)

    # Check every 5 seconds if it is still pushing the box
    frontError = 10
    if g.worldTime % (3*1024) == 0:
        if not ((direction > -frontError and direction < frontError) or \
         (direction < -180+frontError or direction > 180-frontError)):
            #print(f'Could not push to direction {direction}, another robot was blocking')
            changeState(g.State.MOVE_AROUND_OBJECT)

    # Check every 2 seconds if should move around the object
    if g.worldTime % (2*1024) == 0:
        if canSeeGoal():
            #print(f'Can see goal, move around object')
            changeState(g.State.MOVE_AROUND_OBJECT)

File: simulation/src/pusherCommon.cpp
//--------------------------------------------------
// Box Pushing
// pusherCommon.cpp
// Date: 2023-02-08
//--------------------------------------------------
#include "pusherCommon.h"
#include <atta/component/components/rigidBody2D.h>
#include <atta/component/components/transform.h>

void PusherCommon::changeState(PusherComponent* pusher, PusherComponent::State state) {
    // LOG_DEBUG("Pusher" + std::to_string(_entity.getCloneId()), "Changed to state $0", int(state));

    // Don't reset the timer if changed between MOVE_AROUND_OBJECT and PUSH_OBJECT
    if (state == PusherComponent::RANDOM_WALK || pusher->state == PusherComponent::RANDOM_WALK || state == PusherComponent::BE_A_GOAL ||
        pusher->state == PusherComponent::BE_A_GOAL)
        pusher->timer = 0.0f;

    // Change state
    pusher->state = state;
}

void PusherCommon::move(cmp::Entity entity, atta::vec2 direction) {
    constexpr float wheelD = 0.06f; // Wheel distance
    constexpr float wheelR = 0.01f; // Wheel radius
    constexpr float maxPwr = 50.0f; // Motor maximum power (max 0.5m/s)

    auto r = entity.get<cmp::RigidBody2D>();

    // If break motors
    if (direction.x == 0.0f && direction.y == 0.0f) {
        r->setLinearVelocity(atta::vec2(0.0f));
        r->setAngularVelocity(0.0f);
        return;
    }

    // Calculate motor velocities
    direction.normalize();
    float dirAngle = atan2(direction.y, direction.x);                             // Direction angle
    atta::vec2 pwr(cos(dirAngle) - sin(dirAngle), cos(dirAngle) + sin(dirAngle)); // Power
    pwr.normalize();
    pwr *= maxPwr;

    // Calculate linear/angular velocities (differential drive robot)
    float linVel = wheelR / 2.0f * (pwr.x + pwr.y);
    float angVel = wheelR / wheelD * (pwr.x - pwr.y);

    // Apply velocities
    float angle = entity.get<cmp::Transform>()->orientation.get2DAngle();
    r->setLinearVelocity(atta::vec2(linVel * cos(angle), linVel * sin(angle)));
    r->setAngularVelocity(angVel);
}

atta::vec2 PusherCommon::dirToVec(float dir) { return {std::cos(dir), std::sin(dir)}; }

float PusherCommon::distInDirection(const std::array<float, 8>& irs, float dir) {
    int idx = std::round((dir + M_PI) / (2 * M_PI) * 8 - 4);
    if (idx < 0)
        idx = 8 + idx;
    return irs[idx];
}

float angleDistance(float a0, float a1) {
    float dist = a1 > a0 ? (a1 - a0) : (a0 - a1);
    return dist > M_PI ? 2 * M_PI - dist : dist;
}

void PusherCommon::randomWalk(cmp::Entity entity, PusherComponent* pusher, float dt, bool isPaperScript) {
    const float a = 2.0f;
    const float b = 0.2f;
    pusher->randomWalkAux += dt * (rand() / float(RAND_MAX) * a * 2 - a); // Add Unif(-a, a)

    // Clip randomWalkAux angle
    if (pusher->randomWalkAux < -b)
        pusher->randomWalkAux = -b;
    if (pusher->randomWalkAux > b)
        pusher->randomWalkAux = b;
    PusherCommon::move(entity, PusherCommon::dirToVec(pusher->randomWalkAux));

    if (!isPaperScript) {
        // If the goal is no longer visible
        bool noLongerVisible = !pusher->canSeeGoal() && pusher->couldSeeGoal;
        if (noLongerVisible && pusher->angleGreater90) {
            PusherCommon::changeState(pusher, PusherComponent::BE_A_GOAL);
            return;
        }
    }

    // Check if it can see
    bool canSee = pusher->canSeeObject() && (!isPaperScript ? pusher->canSeeGoal() : true);
    if (canSee)
        PusherCommon::changeState(pusher, PusherComponent::APPROACH_OBJECT);
}

void PusherCommon::approachObject(cmp::Entity entity, PusherComponent* pusher, const std::array<float, 8>& irs, bool isPaperScript) {
    const float minCamDist = 0.15f; // Minimum distance to the object to change state (using Camera)
    const float minIrDist = 0.1f;   // Minimum distance to the object to change state (using IR)
    const float minAngle = 0.15f;   // The front angle interval is [-minAngle, minAngle]

    // Check if it can still see
    bool canSee = pusher->canSeeObject() && (!isPaperScript ? pusher->canSeeGoal() : true);
    if (!canSee) {
        PusherCommon::changeState(pusher, PusherComponent::RANDOM_WALK);
        return;
    }

    float dir = pusher->objectDirection;

    // Move to object
    PusherCommon::move(entity, PusherCommon::dirToVec(dir));

    // Check if arrived
    bool isClose = pusher->objectDistance < minCamDist &&                                   // If distance from camera is small
                   PusherCommon::distInDirection(irs, pusher->objectDirection) < minIrDist; // And distance from IR is small
    bool isInFront = (dir < minAngle && dir > -minAngle) || (dir < (-M_PI + minAngle) || dir > (M_PI - minAngle));
    if (isClose && isInFront) {
        if (!pusher->canSeeGoal() && pusher->freeSpaceToPush())
            PusherCommon::changeState(pusher, PusherComponent::PUSH_OBJECT);
        else
            PusherCommon::changeState(pusher, PusherComponent::MOVE_AROUND_OBJECT);
        return;
    }
}

void PusherCommon::moveAroundObject(cmp::Entity entity, PusherComponent* pusher, const std::array<float, 8>& irs, float dt, bool isPaperScript) {
    //----- Check lost object -----//
    if (!pusher->canSeeObject()) {
        PusherCommon::changeState(pusher, PusherComponent::RANDOM_WALK);
        return;
    }

    //----- Check should push -----//
    if (!pusher->canSeeGoal() && pusher->freeSpaceToPush()) {
        PusherCommon::changeState(pusher, PusherComponent::PUSH_OBJECT);
        return;
    }

    //----- Timeout -----//
    // If timer reached zero
    if (pusher->timer >= PusherComponent::pushObjectTimeout) {
        PusherCommon::changeState(pusher, PusherComponent::RANDOM_WALK);
        return;
    }

    //----- Select clockwise/anti-clockwise -----//
    if (!isPaperScript && pusher->canSeeGoal()) {
        float diff = pusher->objectDirection - pusher->goalDirection;
        if (diff >= 2 * M_PI)
            diff -= 2 * M_PI;
        if (diff < 0)
            diff += 2 * M_PI;
        pusher->clockwise = diff >= M_PI;
    }

    //----- Parameters -----//
    atta::vec2 moveVec(1.0f, 0.0f); // Robot move vector (X is to the forward, Y is left)
    const float minDist = 0.1f;
    const float maxDist = 0.25f;

    unsigned idxF = pusher->clockwise ? 0 : 4; // Index at the front of the robot
    if (pusher->objectDirection < 0) {
        // If robot moving "backward", front sensor is behind
        idxF = (idxF == 4) ? 0 : 4;
    }

    //----- Force field -----//
    atta::vec2 objVec = PusherCommon::dirToVec(pusher->objectDirection);

    // Force to move around the object
    moveVec = atta::vec2(-objVec.y, objVec.x);
    if (pusher->clockwise)
        moveVec *= -1;

    // Force to keep distance from object
    if (pusher->objectDistance > 0.2 || PusherCommon::distInDirection(irs, pusher->objectDirection) > 0.2)
        moveVec += objVec;

    //----- Output - move -----//
    PusherCommon::move(entity, moveVec);
}

void PusherCommon::pushObject(cmp::Entity entity, PusherComponent* pusher) {
    // Can't see object anymore
    if (!pusher->canSeeObject()) {
        PusherCommon::changeState(pusher, PusherComponent::RANDOM_WALK);
        return;
    }

    // Can see goal or no free space to push
    if (pusher->canSeeGoal() || !pusher->freeSpaceToPush()) {
        PusherCommon::changeState(pusher, PusherComponent::MOVE_AROUND_OBJECT);
        return;
    }

    // Timeout
    if (pusher->timer >= PusherComponent::pushObjectTimeout) {
        PusherCommon::changeState(pusher, PusherComponent::RANDOM_WALK);
        return;
    }

    // Push
    PusherCommon::move(entity, PusherCommon::dirToVec(pusher->objectDirection));
}

float calcDirection(std::array<cmp::CameraSensor*, 4> cams, unsigned y, Color color) {
    // Create row of colors
    std::array<const uint8_t*, 4> images = {cams[0]->getImage(), cams[1]->getImage(), cams[2]->getImage(), cams[3]->getImage()};
    const unsigned w = cams[0]->width;
    const unsigned h = cams[0]->height;

    // Initialize row to compute
    std::vector<Color> row(w * 4);
    for (unsigned i = 0; i < w * 4; i++) {
        const uint8_t* img = images[i / w];
        unsigned idx = (y * w + (i % w)) * 3;
        Color color(img[idx + 0], img[idx + 1], img[idx + 2]);
        row[i] = color;
    }

    // Calculate intervals
    std::vector<std::pair<int, int>> intervals;
    int start = -1;
    int end = -1;
    for (unsigned i = 0; i < row.size(); i++) {
        if (row[i] != color) {
            if (start != end)
                intervals.push_back({start + 1, end});
            start = i;
            end = i;
        } else
            end = i;
    }
    if (start != end)
        intervals.push_back({start + 1, end});

    // If could not find color, return NaN
    if (intervals.empty()) {
        LOG_WARN("PusherScript", "No interval found when calculating direction");
        return NAN;
    }

    // Merge intervals
    if (intervals.size() > 1 && intervals.front().first == 0 && intervals.back().second == row.size() - 1) {
        intervals.front().first = intervals.back().first;
        intervals.pop_back();
    }

    // Find largest interval
    unsigned idxLargest = -1;
    unsigned sizeLargest = 0;
    for (unsigned i = 0; i < intervals.size(); i++) {
        int start = intervals[i].first;
        int end = intervals[i].second;

        // Calculate size
        unsigned size = 0;
        if (start <= end)
            size = end - start + 1;
        else
            size = (row.size() - start) + end + 1;

        // Update largest
        if (size >= sizeLargest) {
            idxLargest = i;
            sizeLargest = size;
        }
    }

    // Calculate direction to largest interval
    auto interval = intervals[idxLargest];
    // Calculate mean pixel position
    int pixelPos = (interval.first + sizeLargest / 2) % row.size();
    // Convert to [0,1]
    float meanPos = pixelPos / float(row.size());
    // Convert to [-2, 2] (and rotate so 0.0 is to the front)
    meanPos = (meanPos * 4 - 0.5);
    if (meanPos > 2)
        meanPos = -2 + (meanPos - 2);
    // Return direction
    return meanPos * M_PI * 0.5;
}

void PusherCommon::processCameras(PusherComponent* pusher, std::array<cmp::CameraSensor*, 4> cams) {
    PROFILE();

    // If it is not a new image, do not process
    if (cams[0]->captureTime == pusher->lastFrameTime)
        return;

    // Store data about last frame
    pusher->lastFrameTime = cams[0]->captureTime;
    pusher->couldSeeGoal = pusher->canSeeGoal();

    // Initialize values as default
    pusher->objectDirection = NAN;
    pusher->objectDistance = NAN;
    pusher->goalDirection = NAN;
    pusher->goalDistance = NAN;
    pusher->pushDirection = NAN;

    // If there is no image available yet, don't process
    if (cams[0]->captureTime < 0.0f)
        return;

    // Process images
    std::array<const uint8_t*, 4> images = {cams[0]->getImage(), cams[1]->getImage(), cams[2]->getImage(), cams[3]->getImage()};
    const unsigned w = cams[0]->width;
    const unsigned h = cams[0]->height;
    const int startY = h * 0.85;
    for (int y = startY; y >= 0; y--) // Scan from bottom to top (ignore lower pixels where robot is visible)
        for (unsigned x = 0; x < w * 4; x++) {
            // Get pixel values
            const uint8_t* img = images[x / w];
            unsigned idx = (y * w + (x % w)) * 3;
            Color color = {img[idx + 0], img[idx + 1], img[idx + 2]};

            // Update distances
            if (color == objectColor)
                pusher->objectDistance = y / float(h);
            if (color == goalColor)
                pusher->goalDistance = y / float(h);

            // Update goal/object directions
            if (color == objectColor && std::isnan(pusher->objectDirection))
                pusher->objectDirection = calcDirection(cams, y, objectColor);
            if (color == goalColor && std::isnan(pusher->goalDirection))
                pusher->goalDirection = calcDirection(cams, y, goalColor);

            // Update push direction
            if (std::isnan(pusher->pushDirection)) {
                // Get pixel below
                unsigned idxB = ((y + 1) * w + (x % w)) * 3;
                Color colorBelow = {img[idxB + 0], img[idxB + 1], img[idxB + 2]};
                // Check if pixel is object is pixel below is not pusher
                if (color == objectColor && colorBelow != pusherColor && (y == startY || colorBelow != objectColor))
                    pusher->pushDirection = calcDirection(cams, y, objectColor);
            }
        }

    if (pusher->canSeeGoal() && pusher->canSeeObject())
        // Update angle between goal and object greater than 90
        pusher->angleGreater90 = angleDistance(pusher->goalDirection, pusher->objectDirection) > M_PI / 2.0f;
    else if (pusher->canSeeGoal() && !pusher->canSeeObject())
        // Don't do angle check if only goal is visible
        pusher->angleGreater90 = true;
}

